### PR TITLE
issue #36 Degenerate spectrum problem

### PR DESCRIPTION
**Pull request by israelferrazaraujo**

### CODE REVIEW

File: test/test_unitary.py [hunk 1]
```diff
@@ -84,3 +84,38 @@ def test_compute_gates(self):
         calc2 = v_gate @ np.diag(diag).conj().T @ w_gate
         self.assertTrue(np.allclose(calc1, gate1))
         self.assertTrue(np.allclose(calc2, gate2))
+
+    def test_compute_gates_fixed(self):
+        """ test auxiliar funciont compute gates"""
+        # Matrices gate1 and gate2 are not unitary.
```
Comment (adjs) [inline test/test_unitary.py:90]:
```suggestion

```

`qclib.unitary._is_unitary(gate1)` and `qclib.unitary._is_unitary(gate2)` return `True`

File: qclib/unitary.py [hunk 2]
```diff
@@ -87,12 +87,26 @@ def _qsd(gate1, gate2):
...
+
+def _closest_unitary(matrix):
+    svd_u,_,svd_v = np.linalg.svd(matrix)
+    return svd_u.dot(svd_v)
 
 def _compute_gates(gate1, gate2):
 
     d_square, gate_v = np.linalg.eig(gate1 @ gate2.conj().T)
-    list_d = np.sqrt(d_square)
+    list_d = np.sqrt(d_square, dtype=np.complex)
```
Comment (adjs) [inline qclib/unitary.py:103]: I am not sure if we need to include `dtype=np.complex` . If d[k] is real, d[k]**2 >= 0 and if d[k] is complex we don't need to set the type.

File: test/test_unitary.py [hunk 3]
```diff
@@ -84,3 +84,38 @@ def test_compute_gates(self):
...
+                  -2.90528108e-01+2.31698952e-01j, -6.83375688e-01+3.93430693e-01j],
+                 [-5.77541244e-03-1.72752243e-01j,  7.06478710e-01+2.97965199e-02j,
+                   1.76031973e-01-5.86222022e-01j, -2.70215300e-02+3.07814277e-01j]])
+
+        # Let's find the closest unitary matrices to gate1 and gate2.
+        def closest_unitary(matrix):
+            svd_u,_,svd_v = np.linalg.svd(matrix)
+            return svd_u.dot(svd_v)
+        gate1 = closest_unitary(gate1)
+        gate2 = closest_unitary(gate2)
```
Comment (adjs) [inline test/test_unitary.py:115]:
```suggestion

```

This new `_compute_gates` already checks this problem.

File: test/test_baa_schmidt.py [hunk 4]
```diff
@@ -124,6 +124,17 @@ def test_initialize_no_loss(self):
 
         self.assertTrue(np.allclose(state_vector, state))
 
+    def test_initialize_ame(self):
```
Comment (adjs) [inline test/test_baa_schmidt.py:127]:
```suggestion
    def test_initialize_ame(self):
        """ Test initialization of a maximum entangled state"""
```

File: qclib/unitary.py [hunk 5]
```diff
@@ -87,12 +87,26 @@ def _qsd(gate1, gate2):
     circuit = circuit.compose(right_gate, qubits[0:-1])
     return circuit
 
+def _unitary_condition(matrix):
+    return np.conj(matrix.T).dot(matrix)
+
+def _is_unitary(matrix):
+    return np.allclose(_unitary_condition(matrix), np.identity(matrix.shape[0]))
```
Comment (adjs) [inline qclib/unitary.py:94]:
```suggestion

def _is_unitary(matrix):
    is_identity = np.conj(matrix.T).dot(matrix)
    return np.allclose(is_identity, np.identity(matrix.shape[0]))
```

File: test/test_schmidt.py [hunk 6]
```diff
@@ -75,6 +75,17 @@ def test_initialize_full_rank(self):
 
         self.assertTrue(np.allclose(state_vector, state))
 
+    def test_initialize_ame(self):
```
Comment (adjs) [inline test/test_schmidt.py:78]:
```suggestion
    def test_initialize_ame(self):
        """ Test initialization of a maximum entangled state"""
```

**Comment by adjs** [comment]
This PR solves issue #36. Thanks @israelferrazaraujo.

File: qclib/unitary.py (hunk 2 above)
Comment (israelferrazaraujo) [reply qclib/unitary.py:103]: The matrix **gate1 @ gate2.conj().T** is unitary, so the only guarantee we have about the eigenvalues is that their absolute value is 1. There is no restriction on the phase.

File: test/test_unitary.py (hunk 3 above)
Comment (israelferrazaraujo) [reply test/test_unitary.py:114]: I will remove this test. It is no longer necessary.

File: test/test_baa_schmidt.py (hunk 4 above)
Comment (israelferrazaraujo) [reply test/test_baa_schmidt.py:127]: AME = Absolutely Maximally Entangled.

File: test/test_schmidt.py (hunk 6 above)
Comment (israelferrazaraujo) [reply test/test_schmidt.py:78]: AME = Absolutely Maximally Entangled. I will include the comment accordingly.

**Pull request merged**